Add spec for `brew` package URLs (#281)
package-url/purl-spec: PURL-TYPES.rst

      pkg:bitnami/wordpress@6.2.0
      pkg:bitnami/wordpress@6.2.0?arch=arm64

brew
----
``brew`` for Homebrew-based packages:

> - The default repository (tap) is ``homebrew/core``.

Reviewer: **suggestion (non-blocking):** 
```suggestion
- The default repository, which is called a "Tap" in Homebrew terminology, is ``homebrew/core``.
```

> - The default repository (tap) is ``homebrew/core``.

Reviewer 2: I think care needs to be taken here to distinguish the default tap repository `homebrew/core` from the PURL `repository_url` qualifier, which takes a URL. `homebrew/core` is not a valid URL and should not be put into the `repository_url` qualifier. Besides being confusing to have something that isn't a URL in the URL qualifier, this would break any implementation that expects the repository URL to always be a URL.

There should be an example that shows specifying the repository.

Author: > There should be an example that shows specifying the repository.

Might be misunderstanding what you mean, but examples 2 through 4 currently show this:

* `pkg:brew/homebrew/core/sqlite@3.43.2` specifies the repository as `homebrew/core`, which is otherwise the implicit default repo;
* `pkg:brew/some-org/some-tap/some-app@1.2.3` specifies the repository as `some-org/some-tap`, which expands to `https://github.com/some-org/homebrew-some-tap`, per Homebrew's tap expansion rules;
* `pkg:brew/some-org/some-tap/some-app@1.2.3?tap_url=https://git.example.com/some-org/some-tap.git` specifies the repository as `https://git.example.com/some-org/some-tap.git`, which then gets tapped with a custom name of `some-org/some-tap` (which could be anything, since the full URL is specified)

Reviewer 2: I thought the repository would be in a qualifier like how it is with `repository_url`. If it's in the namespace, that should be called out. Normally, a type spec has a line at the top specifying the default repository URL qualifier value, but `brew` doesn't seem to use the repository URL qualifier. I'm trying to avoid people creating PURLs like `pkg:brew/sqlite@3.43.2?repository_url=homebrew/core` instead.

As it's currently written, there's an explanation of how repositories work, but I don't see an explanation of how the repository is encoded into the PURL. Apparently it goes into the namespace, and there are examples of tap names in the namespace, but the namespace field is never mentioned.

Author: > I thought the repository would be in a qualifier like how it is with repository_url. If it's in the namespace, that should be called out.

Gotcha, I'll do this. I'm not super familiar with the idioms here; is the namespace not the right place to put this? A Homebrew Tap is essentially a package namespace, and the purl spec says this about namespaces:

> **namespace**: some name prefix such as a Maven groupid, a Docker image owner,
  a GitHub user or organization. Optional and type-specific.

Reviewer 2: I think namespace should be fine as long as it's explained.

Author: Ah, I think I see the ambiguity here: Homebrew's taps don't really distinguish between the "repository" and the "namespace," which are distinct concepts in purls. For standard/consistent language purposes, I'll refer to the tap only as a namespace.

>   - The tap syntax is ``https://github.com/{org}/Homebrew-{tap}``, so
>     ``homebrew/core`` corresponds to the URL ``https://github.com/homebrew/homebrew-core``.

Reviewer: **suggestion (non-blocking):** 
```suggestion
  - The typical tap identifier expands to the URL ``https://github.com/{org}/Homebrew-{tap}``, so the tap identifier
    ``homebrew/core`` corresponds to the tap URL ``https://github.com/homebrew/homebrew-core``.
```

I'm maybe too wordy here, but for a spec, wordy might be good.

Author: Agreed, wordy is good!

> - The ``name`` is the formula name.

Reviewer: **suggestion (non-blocking):** Add this within the previous list item:
```suggestion
  - Any git URL is a valid tap URL (see ``tap_url``)
- The ``name`` is the formula name.
```

This clarifies from the start, but maybe the `tap_url` bullet belongs in here?

> - The ``version`` is the formula version.

Reviewer: **question (non-blocking):** When thinking about the examples below, I think you should clarify  how purl spec should handle version-in-formula formulae.

As in, if I want PostgreSQL 12.17 (current release as of this posting), should the purl be

    pkg:brew/postgresql@12@12.17

or 

    pkg:brew/postgresql@12.17

and let whatever's reading the purl — Homebrew, ostensibly — figure out how that version maps to our major-version formulae?

I speculate only the latter is a valid purl, a safe assumption because I don't think it's a good idea to allow two `@` in the purl.

Author: Thanks for calling this out -- my read of the purl spec is that the name and other components are always URL-escaped, e.g. `pkg:brew/postgresql@12@12.17` would actually be encoded as `pkg:brew/postgresql%4012@12.17` in a "wire format" context. So the purl *shoudl* always unambiguously map to a package, without the end client having to do additional disambiguation work.

This is a readability sacrifice, but my (potentially wrong) understanding is that purls are mostly meant to show up only in machine-readable contexts anyways.

Reviewer: Sigh, it's not pretty, but if that's the way it needs to be, so be it. If that's the case, I'd recommend adding an example with that…

Author: Done! Looks like the other examples for other ecosystem purls also contain URL-escaping examples, so we're in-line here 🙂

Author: Yep, I added @colindean's example below. Adding an explicit callout also makes sense to me; I'll add some language for that.

- Qualifier ``tap_url``: for taps that are not on GitHub or otherwise require an explicit URL,

>   this is the full URL to the tap.

Reviewer: **question (non-blocking):** Should there be another parameter to name the tap? We use the named tap at work, e.g.

    brew tap myco/brewhouse git@git.myco.com:brew/house.git

and the formula, uh, `myco-ctl` could have a purl

    pkg:brew/myco/brewhouse/myco-ctl?tap_url=git@git.myco.com:brew/house.git

and brew would know how to name the tap if it needed to be tapped, or perhaps

    pkg:brew/myco/brewhouse/myco-ctl?tap_url=git@git.myco.com:brew/house.git&tap_name=myco/brewhouse

or more simply

    pkg:brew/myco-ctl?tap_url=git@git.myco.com:brew/house.git&tap_name=myco/brewhouse

Author: My 0.02c would be for your first form:

```
pkg:brew/myco/brewhouse/myco-ctl?tap_url=git@git.myco.com:brew/house.git
```

I _think_ that's unambiguous: the `github.com/{org}/homebrew-{tap}` logic only applies when a `tap_url` isn't present, so I think this is the shortest form that conveys all needed information. But maybe I've overlooked something?

- Examples::

      pkg:brew/sqlite@3.43.2
      pkg:brew/homebrew/core/sqlite@3.43.2
      pkg:brew/some-org/some-tap/some-app@1.2.3

>       pkg:brew/some-org/some-tap/some-app@1.2.3&tap_url=https://git.example.com/some-org/some-tap.git

Reviewer: **nitpick (blocking):** 
```suggestion
      pkg:brew/some-org/some-tap/some-app@1.2.3?tap_url=https://git.example.com/some-org/some-tap.git
```

      pkg:bitnami/wordpress@6.2.0
      pkg:bitnami/wordpress@6.2.0?arch=arm64

brew
----
``brew`` for Homebrew-based packages:

- The default repository, which is called a "Tap" in Homebrew terminology, is ``homebrew/core``.

  - The typical tap identifier expands to the URL ``https://github.com/{org}/Homebrew-{tap}``, so
    the tap identifier ``homebrew/core`` corresponds to the tap URL
    ``https://github.com/homebrew/homebrew-core``.
  - Any git URL is a valid tap URL (see ``tap_url``)

- The ``name`` is the formula name.
- The ``version`` is the formula version.
- Qualifier ``tap_url``: for taps that are not on GitHub or otherwise require an explicit URL,
  this is the full URL to the tap.
- Examples::

>       pkg:brew/sqlite@3.43.2

Reviewer: **suggestion:** Based on the encoding comment above:
```suggestion
      pkg:brew/sqlite@3.43.2
      pkg:brew/postgresql%4012@12.17
```

      pkg:bitnami/wordpress@6.2.0?arch=arm64&distro=debian-12
      pkg:bitnami/wordpress@6.2.0?arch=arm64&distro=photon-4

brew
----
``brew`` for Homebrew-based packages:

> - There is no default package repository; this should be implied either from

Reviewer 3: > Do we have examples of other public homebrew repos BTW?

https://github.com/CycloneDX/homebrew-cyclonedx

> - There is no default package repository; this should be implied either from

Reviewer 4: For a standard installation in Europe for example there is a default of `Homebrew/core` and `Homebrew/cask` which are backed by an API. Behind the Great Firewall people are using mirrors of the API and packages. But people can also opt to use the GitHub repository as a backing for core and cask.

I believe the API can also work for other repositories, but I'm not 100% sure about that. They might need to have a Git repository.

> - There is no default package repository; this should be implied either from

Reviewer 3: > > nope. homebrew/cask is not enabled by default. it might be auto-enabled when needed.
> 
> As a homebrew maintainer I can most definitely assure you it is, maybe you tried Homebrew on Linux where not all defaults line up with the main macOS platform?

https://docs.brew.sh/Taps makes the impression that [homebrew/core](https://github.com/Homebrew/homebrew-core) is the default repo. 

> If your tap contains a formula that is also present in [homebrew/core](https://github.com/Homebrew/homebrew-core), that’s fine, but you would need to specify its fully qualified name in the form <user>/<repo>/<formula> to install your version.

and again: even the repo's description states that it is the default.

> - There is no default package repository; this should be implied either from

Reviewer 3: casks are only instated from a certain repo since it was auto-enabled on macOS the name did not exist in the actual default repo.
again: [homebrew/core](https://github.com/Homebrew/homebrew-core) is default repo, right?

the tap docs say so (see above), the repo description says so (see above), 
and also the man page say so: https://github.com/Homebrew/brew/blob/fe4f472c8d55e336ff20a46f91194e2f0d8b1228/manpages/brew.1#L1597
code says so: https://github.com/Homebrew/brew/blob/fe4f472c8d55e336ff20a46f91194e2f0d8b1228/Library/Homebrew/api/formula.rb#L33

> - There is no default package repository; this should be implied either from

Author: No, casks are not a special kind of formula. They're a completely different kind of package entirely: formulae are for builds of open source projects; casks are for application distributions where Homebrew can't provide a build for licensing or logistical reasons.

Homebrew supports two different things, both of which are defaults. If you do `brew install foo` you will get either a formula or a cask, depending on what `foo` is. This is because both homebrew-core and homebrew-cask are enabled by default. So it's not accurate to say that homebrew-core is *the* default repo; it's *one* of two default repos.

I think maybe the documentation is confusing here: the tap documentation is mostly about formulae, but casks share some of the same machinery. The basic version is that `homebrew-cask` is 100% a default repository, and `brew install foo` will install the foo cask by default on macOS with no user changes or configuration needed.

> - There is no default package repository; this should be implied either from

Reviewer 3: in case it is possible to distinguish casks from formulas in any way, i'd propose:

```rst
- the ``namespace`` is the package type - valid values: 
  - "formula"
  - "cask"
- The ``name`` is the formula name. Formula names that contain ``@`` must be percent-encoded, such as ``postgresql%4012`` for ``postgres@12``.
- The ``version`` is the formula version.
- Qualifier ``repository_url``: URL to the tap. 
  - Default for namespace "formula" is ``https://github.com/Homebrew/homebrew-core.git``. 
  - Default for namespace "cask" is ``https://github.com/Homebrew/homebrew-cask.git``. 
- Examples::
      pkg:brew/formula/sqlite@3.43.2
      pkg:brew/formula/postgresql%4012@12.17
      pkg:brew/cask/some-app@1.2.3
      pkg:brew/cask/some-app@1.2.3?repository_url=https://git.example.com/some-org/some-tap.git
```

> - There is no default package repository; this should be implied either from

Reviewer 4: There is a technical way to distinguish between them, but users don't. They install docker and rarely care if it's built from source or not
